src/librustc/mir/repr.rs

```diff
@@ -614,12 +618,60 @@ impl<'tcx> Debug for Lvalue<'tcx> {
 }
 
 ///////////////////////////////////////////////////////////////////////////
+// Scopes
+
+#[derive(Clone, Debug, RustcEncodable, RustcDecodable)]
+pub struct ScopeDataVec {
+    pub vec: Vec<ScopeData>
+}
+
+impl ScopeDataVec {
+    pub fn new() -> Self {
+        ScopeDataVec { vec: Vec::new() }
+    }
+}
+
+impl Index<ScopeId> for ScopeDataVec {
+    type Output = ScopeData;
+
+    #[inline]
+    fn index(&self, index: ScopeId) -> &ScopeData {
+        &self.vec[index.index()]
+    }
+}
+
+impl IndexMut<ScopeId> for ScopeDataVec {
+    #[inline]
+    fn index_mut(&mut self, index: ScopeId) -> &mut ScopeData {
+        &mut self.vec[index.index()]
+    }
+}
+
+#[derive(Copy, Clone, Debug, RustcEncodable, RustcDecodable)]
+pub struct ScopeId(u32);
```

Comment:
I've been wanting for some time to make a macro for this purpose.

On Tue, Mar 22, 2016 at 09:36:17AM -0700, Eduard-Mihai Burtescu wrote:
> > +
> > +    #[inline]
> > +    fn index(&self, index: ScopeId) -> &ScopeData {
> > +        &self.vec[index.index()]
> > +    }
> > +}
> > +
> > +impl IndexMut<ScopeId> for ScopeDataVec {
> > +    #[inline]
> > +    fn index_mut(&mut self, index: ScopeId) -> &mut ScopeData {
> > +        &mut self.vec[index.index()]
> > +    }
> > +}
> > +
> > +#[derive(Copy, Clone, Debug, RustcEncodable, RustcDecodable)]
> > +pub struct ScopeId(u32);
> 
> I *almost* feel like we should have a general-purpose `Id<T>` type for these kinds of things.
> 
> ---
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/rust-lang/rust/pull/32428/files/9e8bcf4901c547a29ceed5e27606110ff26a9842#r57020384